### ISSUE TITLE
Generated Paths_ module conflicts with classy-prelude

### ISSUE THREAD

**Issue by 23Skidoo** [haskell/cabal#5962]
I have the following common stanza:

```
common classy-prelude
  build-depends:       base >=4.11 && <4.13
                     , classy-prelude
  mixins:              base hiding (Prelude)
                     , classy-prelude (ClassyPrelude as Prelude)
```

Unfortunately, the autogenerated `Paths_` module conflicts with `classy-prelude` because they both define `catchIO`:

```
    Ambiguous occurrence ‘catchIO’
    It could refer to either ‘Prelude.catchIO’,
                             imported from ‘Prelude’ at .../autogen/Paths_[...].hs:13:1-14
                             (and originally defined in ‘unliftio-0.2.9.0:UnliftIO.Exception’)
                          or ‘Paths_[...].catchIO’,
                             defined at .../autogen/Paths_[...].hs:26:1
```

**Comment by hvr** [comment]
I wouldn't call this a bug. The `Paths_*` module can't operate completely agnostic to `build-depends`; it has a couple of basic requirements as to the environment it's compiled in. So I'd rather consider this a documentation issue than a bug.

**Comment by 23Skidoo** [comment]
We can easily fix it with an explicit import list.

**Comment by phadej** [comment]
We can, but we need still to document most assumptions

- A different prelude can redefine `type IO a = HasCallStack => GHC.IO a`...
- or makes `getEnv` take Text or ByteString or
- makes `FilePath /= String`

And these aren't even too far fetched changes.

**Comment by 23Skidoo** [comment]
True.

**Comment by 23Skidoo** [comment]
Related: #1170.

**Comment by matthewess** [comment]
Is there a good workaround for this? I'm using [RIO](https://github.com/commercialhaskell/rio) as a prelude replacement and running into the same issue. The only obvious way I can see to fix for myself as of now is to `import RIO hiding (catchIO)`

**Comment by rossabaker** [comment]
#1170 looks like a harder problem, but given the assumption of a standard base, I think it's worthwhile progress if we can insulate ourselves from alternate preludes.  This compiles for me, and I'd be happy to PR it if there's interest:

```haskell
{-# LANGUAGE NoImplicitPrelude #-}

-- ...

import qualified Control.Exception as Exception
import Control.Monad (return)
import Data.List ((++))
import Data.Version (Version(..))
import System.Environment (getEnv)
import System.IO as IO
```

**Comment by hvr** [comment]
@rossabaker except it doesn't actually shield you from alternative preludes at all... ;-)

**Comment by adamConnerSax** [comment]
I know this is an old issue, but when migrating from Cabal-3.4 to Cabal-3.6  and using relude:

```Haskell
 mixins:  base hiding (Prelude)
         , relude (Relude as Prelude
         , Relude.Extra)
```

which then generates an error when compiling the paths module because relude defines ```head``` to operate on ```NonEmpty``` rather than ```[]```:

```haskell
XXX/dist-newstyle/build/x86_64-osx/ghc-8.10.7/knit-haskell-0.10.1.0/build/autogen/Paths_knit_haskell.hs:66:22: error:
    • 'last' works with 'NonEmpty', not ordinary lists.
      Possible fix:
          Replace: [Char]
          With:    NonEmpty Char
      
      However, you can use 'last' with the ordinary lists.
      Apply 'viaNonEmpty' function from relude:
          viaNonEmpty last (yourList)
      Note, that this will return 'Maybe Char'
      therefore it is a safe function unlike 'last' from the standard Prelude
    • In the first argument of ‘isPathSeparator’, namely ‘(last dir)’
      In the expression: isPathSeparator (last dir)
      In a stmt of a pattern guard for
                     an equation for ‘joinFileName’:
        isPathSeparator (last dir)
   |
66 |   | isPathSeparator (last dir) = dir ++ fname
   |                      ^^^^^^^^
```

Going back to 3.4 resolves this.

Can someone explain how to work around this in more detail?  Like what is meant by "We can easily fix it with an explicit import list"?  Where would the explicit import list go?

I can just go back to cabal 3.4 for now but that doesn't seem like a long-term solution.

**Comment by fishtreesugar** [comment]
I think it already fixed in https://github.com/haskell/cabal/pull/7650, and cabal-install 3.6.2 will release in recent days https://github.com/haskell/cabal/issues/7671

**Comment by jneira** [comment]
The concrete issue was fixed by #7650 and we have #7761 to rack a more robust solution so i am gonna close this if you dont mind
